### ISSUE TITLE
cilium pods keep at pending status

### ISSUE THREAD

**Issue by purplepalmdash** [admiraltyio/multicluster-scheduler#50]
Followed the `README.md` for setting up multicluster-scheduler with two kind clusters(cilium-1.8.1 as their CNI, no cluster mesh ),  virtual nodes have been created in cluster1, but cilium pods keep at pending status... how to solve this problem?

ps: everingthing is ok when using kind's default cni("kindnetd")

```
# kubectl --context "$CLUSTER1" get node
NAME                     STATUS   ROLES     AGE   VERSION
admiralty-c1             Ready    cluster   21m
admiralty-c2             Ready    cluster   21m
cluster1-control-plane   Ready    master    85m   v1.18.2
cluster1-worker          Ready    <none>    84m   v1.18.2

# kubectl --context "$CLUSTER2" get node
NAME                     STATUS   ROLES    AGE   VERSION
cluster2-control-plane   Ready    master   83m   v1.18.2
cluster2-worker          Ready    <none>   83m   v1.18.2

# kubectl --context "$CLUSTER1" get node -l virtual-kubelet.io/provider=admiralty
NAME           STATUS   ROLES     AGE   VERSION
admiralty-c1   Ready    cluster   20m
admiralty-c2   Ready    cluster   20m

```
Pending pods: 
```
# kubectl --context "$CLUSTER1" get pods --all-namespaces -o wide | grep -i pending
kube-system          cilium-6mrvl                                                  0/1     Pending   0          22m   <none>       admiralty-c1             <none>           <none>
kube-system          cilium-node-init-7jsvq                                        0/1     Pending   0          22m   <none>       admiralty-c2             <none>           <none>
kube-system          cilium-node-init-s89xs                                        0/1     Pending   0          22m   <none>       admiralty-c1             <none>           <none>
kube-system          cilium-vsnnh                                                  0/1     Pending   0          22m   <none>       admiralty-c2             <none>           <none>
# kubectl --context "$CLUSTER1" describe po cilium-6mrvl -n kube-system
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  32m   default-scheduler  0/4 nodes are available: 1 Insufficient cpu, 1 Insufficient memory, 1 Too many pods, 3 node(s) didn't match node selector.
  Normal   Scheduled         32m   default-scheduler  Successfully assigned kube-system/cilium-6mrvl to admiralty-c1

```

**Comment by adrienjt** [comment]
Cilium's two DaemonSets have the broadest tolerations:
```yaml
spec:
  template:
    spec:
      tolerations:
      - operator: Exists
```
which means "tolerate all taints", but they should not extend to virtual nodes.
The only way to restrict the toleration, i.e., to say "tolerate all taints except for one" is to add a node anti-affinity:
```yaml
spec:
  template:
    spec:
      tolerations:
      - operator: Exists
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: virtual-kubelet.io/provider
                operator: DoesNotExist
```
Please let me know if that works. This should be added to our readme.

**Comment by purplepalmdash** [comment]
Thanks. Added the `nodeAffinity` I could make multicluster-scheduler working together with cilium.   

But another issue is `coredns` pods remains creating and stays at `Pending` status:    

```
root@cilium-1:~# kubectl get pods --all-namespaces | grep -i coredns | grep -i running
kube-system    coredns-77b5f79dbf-74zd9                                      1/1     Running             0          4m56s
kube-system    coredns-77b5f79dbf-r89bl                                      1/1     Running             0          4m56s
root@cilium-1:~# kubectl get pods --all-namespaces | grep -i coredns | grep -i pending | wc -l
780
root@cilium-1:~# kubectl get rs -n kube-system
NAME                                    DESIRED   CURRENT   READY   AGE
cilium-etcd-operator-584788b99c         1         1         1       62m
cilium-operator-7cd598bdf6              1         1         1       62m
coredns-77b5f79dbf                      782       782       2       11m
dns-autoscaler-85f898cd5c               1         1         1       75m
etcd-operator-59cf4cfb7c                1         1         1       18m
kubernetes-dashboard-857df7d6f7         1         1         1       75m
```
describe a `Pending` pod:    

```
root@cilium-1:~# kubectl describe po coredns-77b5f79dbf-ztz9v -n kube-system
Events:
  Type     Reason            Age                   From               Message
  ----     ------            ----                  ----               -------
  Warning  FailedScheduling  2m27s (x3 over 8m1s)  default-scheduler  0/4 nodes are available: 2 node(s) didn't match node selector, 2 node(s) didn't match pod affinity/anti-affinity.
```
Adding nodeAffinity seems doesn't works for coredns:   

```
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: virtual-kubelet.io/provider
                operator: DoesNotExist
          preferredDuringSchedulingIgnoredDuringExecution:
          - preference:
              matchExpressions:
              - key: node-role.kubernetes.io/master
                operator: In
                values:
                - ""
            weight: 100

```

**Comment by purplepalmdash** [comment]
By editing  `"max":4` to `dns-autoscaler`'s configmap I could shrink the coredns pods to `4`(2 pending, 2 running), 

```
# kubectl edit configmap dns-autoscaler -n kube-system
apiVersion: v1
data:
  linear: '{"coresPerReplica":256,"min":2,"max":4,"nodesPerReplica":16,"preventSinglePointFailure":true}'
kind: ConfigMap
metadata:
  creationTimestamp: "2020-07-23T08:10:19Z"
  name: dns-autoscaler
  namespace: kube-system
  resourceVersion: "52908"
  selfLink: /api/v1/namespaces/kube-system/configmaps/dns-autoscaler
  uid: 4914ed87-d754-425b-bd49-cbb44f2ea6e4

```
coredns pods:   
```
# kubectl get pods --all-namespaces -o wide | grep -i coredns
kube-system    coredns-77b5f79dbf-74zd9                                      1/1     Running   0          39m     10.233.64.204   mouse-1        <none>           <none>
kube-system    coredns-77b5f79dbf-d6vm8                                      0/1     Pending   0          39m     <none>          <none>         <none>           <none>
kube-system    coredns-77b5f79dbf-m5bd4                                      0/1     Pending   0          39m     <none>          <none>         <none>           <none>
kube-system    coredns-77b5f79dbf-r89bl                                      1/1     Running   0          39m     10.233.65.73    mouse-2        <none>           <none>
```

dns-autoscaler auto-caculate the max coredns pod number to be `782`.  I think we should delete the dns-autoscaler when using multicluster-scheduler

**Comment by adrienjt** [comment]
Please refer to #43, fixed in v0.10. Indeed, you can either disable DNS autoscaling, or install multicluster-scheduler v0.10.0-rc.1 (which reminds me we should release v0.10.0 proper, since the release candidate has baked long enough).
